lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -721,6 +723,498 @@ struct GrandCentralPass : public GrandCentralBase<GrandCentralPass> {
+      NamedAttrList attrs;
+      if (auto maybeDescription = field.get("description"))
+        attrs.append("description", maybeDescription.cast<StringAttr>());
+      attrs.append("name", name);
+      attrs.append("tpe", tpe.getAs<StringAttr>("class"));
+      elements.push_back(eltAttr.getValue());
+    }
+    // Add an annotation that stores information necessary to construct the
+    // module for the view.  This needs the name of the module (defName) and the
+    // names of the components inside it.
+    NamedAttrList attrs;
+    attrs.append("class", classAttr);
+    attrs.append("defName", defName);
+    if (description)
+      attrs.append("description", description.getValue());
+    attrs.append("elements", ArrayAttr::get(context, elements));
+    if (id)
+      attrs.append("id", id.getValue());
+    attrs.append("name", name);
+    return DictionaryAttr::getWithSorted(context, attrs);
+  }
+
+  // An AugmentedGroundType looks like:
+  //   "ref": ReferenceTarget
+  //   "tpe": GroundType
+  // The ReferenceTarget is not serialized to a string.  The GroundType will
+  // either be an actual FIRRTL ground type or a GrandCentral uninferred type.
+  // This can be ignored for us.
+  if (classBase == "GroundType") {
+    auto maybeTarget = refToTarget(augmentedType.getAs<DictionaryAttr>("ref"));
+    if (!maybeTarget) {
+      mlir::emitError(loc, "Failed to parse ReferenceTarget").attachNote()
+          << "See the full Annotation here: " << root;
+      return None;
+    }
+
+    auto id = state.newID();
+
+    // TODO: We don't support non-local annotations, so force this annotation
+    // into a local annotation.  This does not properly check that the
+    // non-local and local targets are totally equivalent.
+    auto target = maybeTarget.getValue();
+
+    NamedAttrList elementIface, elementScattered, dontTouch;
+
+    // Generate annotations for the ground type.  When possible, generate a dead
+    // wire "tap" on the actual ground type to not block optimizations.  The
+    // following cases can occur:
+    //
+    // 1. The tap is on an external module port.
+    // 2. The tap is on a module port.
+    // 3. The tap is on something inside a module.
+    //
+    // The case of (1) is handled WITHOUT a tap and the ground type stays on the
+    // port.  The cases of (2) and (3) are handled the same: a tap is generated,
+    // the tap is driven by the ground type, and any annotations are placed on
+    // the tap.  (2) requires port-specific handling logic.  (Also, remember
+    // that it is okay to have an output port on the RHS of a connect.  This
+    // avoids having to special case input and output ports.)
+    //
+    // When a ground type tap is on an aggregate type, then one tap of the full
+    // aggregate is generated.  The annotation is then only applied to the
+    // ground type.  This tap will be re-used if other ground types reference
+    // this same tap.
```

Comment:
I think this needs to be changed. I'm running this branch on some internal designs and seeing situations where there are a lot of extra wires getting created (and actually extra ports, too 😂 ). I think this is due to generating a full copy of the aggregate which has one leaf tapped and then the whole aggregate tap is being kept around unnecessarily.

Reply:
I removed this in favor of generating one wire tap for each leaf. This has the effect of producing a good amount of port removal in an internal design. Added `_gctTap*` wires do show up in the final design. However, we can improve this in the future.